ebu_tt_live/bindings/__init__.py

```diff
@@ -580,24 +580,29 @@ class LiveStyledElementMixin(StyledElementMixin):
 # removed from version 1.0 of the specification.
 
 class RoleMixin(object):
-    _computed_roles = None
+    _computed_roles = set()
+
     def _semantic_compute_roles(self, dataset):
+        current_roles = set()
         if self.role is not None:
-            self._computed_roles = self.role
-        else:
-            if dataset.get('metadata_roles') is not None and len(dataset['metadata_roles']) > 0:
-                self._computed_roles = dataset['metadata_roles'][-1]
-            else:
-                self._computed_roles = None
+            current_roles.update(self.role)
+
+        if dataset.get('role_stack') and dataset['role_stack']:
+            current_roles.update(dataset['role_stack'][-1])
+
+        self._computed_roles = current_roles
 
     def _semantic_push_computed_roles(self, dataset):
-        if dataset.get('metadata_roles') is None:
-            dataset['metadata_roles'] = []
-        dataset['metadata_roles'].append(self._computed_roles)
+        if dataset.get('role_stack') is None:
+            dataset['role_stack'] = []
+        dataset['role_stack'].append(self._computed_roles)
 
     def _semantic_pop_computed_roles(self, dataset):
-        if dataset.get('metadata_roles') is not None and len(dataset['metadata_roles']) > 0:
-            dataset['metadata_roles'].pop()
+        if dataset.get('role_stack') and len(dataset['role_stack']) > 0:
+            dataset['role_stack'].pop()
+        if dataset.get('metadata_roles') and len(dataset['metadata_roles']) > 0:
+            self._computed_roles.update(dataset['metadata_roles'])
+            dataset['metadata_roles'].clear()
```

Comment:
Feels weird to put this into the `pop` method - I think it should be in the `_semantic_compute_roles` method but we should be careful when we call it.

Reply:
I recalled we need to clear in after traversal, that's why I have added it here

Reply:
ok I think I misunderstood, now fixed it